kafka/record/legacy_records.py

```diff
@@ -0,0 +1,413 @@
+            " key={!r}, value={!r}, crc={!r})".format(
+                self._offset, self._timestamp, self._timestamp_type,
+                self._key, self._value, self._crc)
+        )
+
+
+class LegacyRecordBatchBuilder(ABCRecordBatchBuilder, LegacyRecordBase):
+
+    def __init__(self, magic, compression_type, buffer=None):
+        self._magic = magic
+        self._compression_type = compression_type
+        self._buffer = buffer if buffer is not None else io.BytesIO()
+
+    def append(self, offset, timestamp, key, value, headers=None):
+        """ Append message to batch.
+        """
+        assert not headers, "Headers not supported in v0/v1"
+        size = self.size_in_bytes(offset, timestamp, key, value)
+        msg_buffer = bytearray(size)
+        crc = self._encode_msg(msg_buffer, offset, timestamp, key, value)
+        self._buffer.write(msg_buffer)
+        return crc, size
+
+    def _encode_msg(self, msg_buffer, offset, timestamp, key, value,
+                    attributes=0):
+        """ Encode msg data into the `msg_buffer`, which should be allocated
+            to at least the size of this message.
+        """
+        magic = self._magic
+
+        # Write key and value
+        pos = self.KEY_OFFSET_V0 if magic == 0 else self.KEY_OFFSET_V1
+
+        if key is None:
+            struct.pack_into(">i", msg_buffer, pos, -1)
+            pos += self.KEY_LENGTH
+        else:
+            key_size = len(key)
+            struct.pack_into(">i", msg_buffer, pos, key_size)
+            pos += self.KEY_LENGTH
+            msg_buffer[pos: pos + key_size] = key
+            pos += key_size
+
+        if value is None:
+            struct.pack_into(">i", msg_buffer, pos, -1)
+            pos += self.VALUE_LENGTH
+        else:
+            value_size = len(value)
+            struct.pack_into(">i", msg_buffer, pos, value_size)
+            pos += self.VALUE_LENGTH
+            msg_buffer[pos: pos + value_size] = value
+            pos += value_size
+        length = pos - self.LOG_OVERHEAD
+
+        # Write msg header. Note, that Crc will be updated later
+        if magic == 0:
+            self.HEADER_STRUCT_V0.pack_into(
+                msg_buffer, 0,
+                offset, length, 0, magic, attributes)
+        else:
+            self.HEADER_STRUCT_V1.pack_into(
+                msg_buffer, 0,
+                offset, length, 0, magic, attributes, timestamp)
+
+        # Calculate CRC for msg
+        crc_data = memoryview(msg_buffer)[self.MAGIC_OFFSET:]
+        crc = calc_crc32(crc_data)
+        struct.pack_into(">I", msg_buffer, self.LOG_OVERHEAD, crc)
+        return crc
+
+    def _maybe_compress(self):
+        if self._compression_type:
+            data = self._buffer.getvalue()
+            if self._compression_type == self.CODEC_GZIP:
+                compressed = gzip_encode(data)
+            elif self._compression_type == self.CODEC_SNAPPY:
+                compressed = snappy_encode(data)
```

Comment:
When I test locally and snappy libraries are not available, this will raise a `NotImplementedError` and crash the producer IO thread:
```
ERROR:kafka.producer.sender:Uncaught error in kafka producer I/O thread
```
But this is only logged and producer.send() still returns a future. If you try to resolve the future with `.get()` it hangs forever. We should probably improve this so that the exception is raised to the user or at least does not crash the IO thread and is used to resolve the future. My preference is the former.